File: python/3/source.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-
'''
Created on 2015-1-7
@author: Ben
'''

class Solution:
    # @return an integer
    def lengthOfLongestSubstring(self, s):
        li = []
        len_max = 0
        for i in range(len(s)):
            if not s[i:i+1] in li:
                li.append(s[i:i+1])
                # 如果到字符串结尾：
                if i == len(s)-1:
                    len_cur = len(li)
                    len_max = len_cur if len_cur > len_max else len_max 
            else:
                len_cur = len(li)
                len_max = len_cur if len_cur > len_max else len_max
                # 更新li里面的值，出现重复值之前(包含重复值)的全部去除
                # 得到出现重复值的索引
                index = li.index(s[i:i+1])
                # 注意，每次去除li里面第一个元素，因为去除第一个元素后，更新后的li的一个元素是之前li的第二个元素
                for j in range(index+1):
                    li.remove(li[0])
                li.append(s[i:i+1])
        return len_max
    
so = Solution()
print(so.lengthOfLongestSubstring('abcabc'))
# print(so.lengthOfLongestSubstring('wlrbbmqbhcdarzowkkyhiddqscdxrjmowfrxsjybldbefsarcbynecdyggxxpklorellnmpapqfwkhopkmco'))

File: c++/12/source.cpp
/**
 * @author piaoyidage
 * @version 创建时间：2015-06-09
 *
 */	


class Solution {
public:
	string intToRoman(int num) {
		// 初始化
		string roman = "";
		
		if (num == 1000)
		{
			roman = "M";
		}
		if (num == 100)
		{
			roman = "C";
		}
		if (num == 10)
		{
			roman = "X";
		}

		// 如果数字大于1000
		if (num > 1000)
		{
			int indexFourth = num / 1000;
			for (int i = 0; i < indexFourth; i++)
			{
				roman += "M";
			}
			roman += intToRoman(num-indexFourth*1000);
		}

	    // 大于100小于100
		if(num > 100 && num < 1000)
		{
			roman = range(num, 100, "C", "D", "M");
		}
        // 大于10小于100
		if (num > 10 && num < 100)
		{
			roman = range(num, 10, "X", "L", "C");
		}
        // 大于0小于10
		if (num > 0 && num < 10)
		{
			roman = range(num, 1, "I", "V", "X");
		}
		
		return roman;
	}
    
    // 根据num值，以及区间、当前区间的罗马标识符，表示当前数字
	string range(int num, int begin, string a, string b, string c)
	{
		string roman = "";

		int index = num / begin;
		switch (index)
		{
		case 1:
		case 2:
		case 3:
			for (int i = 0; i < index; i++)
			{
				roman += a;
			}
			break;
		case 4:
			roman += a;
			roman += b;
			break;
		case 5:
			roman += b;
			break;
		case 6:
		case 7:
		case 8:
			roman += b;
			for (int i = 0; i < index - 5; i++)
			{
				roman += a;
			}
			break;
		case 9:
			roman += a;
			roman += c;
			break;
		}
		roman += intToRoman(num-index*begin);

		return roman;
	}
};